### PR TITLE
Infer Required Props from PropTypes

### PR DESCRIPTION
**Pull request by chrisgarber**
This should do it

### CODE REVIEW

**Comment by chrisgarber** [comment]
Any chance you could pull this into personal/bcherny/61 and merge into master? We would like to start using if possible

**Comment by chrisgarber** [comment]
Hey, I understand if you're busy, but if there's a reason you don't want to merge this in I would appreciate knowing what it is so I can fix the issue.

File: index.tsx [hunk 1]
```diff
@@ -58,10 +59,10 @@ function normalizeBindingNames<Props>(
     return fromPairs(bindingNames.map(_ => [_, '<'])) as AngularBindings<Props>
   }
   if (bindingNames) {
-    return mapValues(bindingNames, _ => _.optional ? '<?' : '<')
+    return mapValues(bindingNames, (_: {optional: boolean}) => _.optional ? '<?' : '<') as AngularBindings<Props>
```
Comment (bcherny) [inline index.tsx]: Do you really need to manually type `_`?

File: index.tsx [hunk 2]
```diff
@@ -58,10 +59,10 @@ function normalizeBindingNames<Props>(
...
   }
   if (Class.propTypes) {
-    return mapValues(Class.propTypes, () => '<' as '<')
+    return mapValues(Class.propTypes as {[K in keyof Props]:PropTypes.Requireable<any>}, (_:PropTypes.Requireable<any>) => !_.isRequired ? '<' : '<?') as AngularBindings<Props>
```
Comment (bcherny) [inline index.tsx]: Shouldn't this be the opposite? If `_.isRequired` then we generate `<`, otherwise we generate `<?`.

File: index.tsx [hunk 3]
```diff
@@ -2,6 +2,7 @@ import { IAugmentedJQuery, IComponentOptions } from 'angular'
 import fromPairs = require('lodash.frompairs')
 import mapValues = require('lodash.mapvalues')
 import NgComponent from 'ngcomponent'
+import * as PropTypes from 'prop-types'
```
Comment (bcherny) [inline index.tsx]: Is there a way we can avoid this dependency?

File: CHANGES.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+CHANGES
+
+allow using proptypes to define optionality
```
Comment (bcherny) [inline CHANGES.txt]: Please rm this file.

File: index.tsx (hunk 2 above)
Comment (chrisgarber) [reply index.tsx]: Yes, this is the most counter intuitive bit, _.isRequired is the prop-types object that defines the required version of the type. If that object exists then the current type is not required. The tests should back this up.

File: index.tsx (hunk 3 above)
Comment (chrisgarber) [reply index.tsx]: we're only truly dependent on the types here, we don't use any of the actual code. In order to use the Requireable type on line 65 (which I needed to make the code compile) we need to import this typing. If we can reference the type without importing the library I'm 100% down. Perhaps a reference path to the @types relative dir?

File: index.tsx (hunk 1 above)
Comment (chrisgarber) [reply index.tsx]: I've been weird different typing issues as though my typescript version is different or something. If you remove this and it compiles for you I'm not at all attached to this.

**Comment by chrisgarber** [comment]
No worries about the delay, I just wanted to make sure I didn't lose you. I understand that maintaining libraries can be a huge time suck and your support is appreciated. I've responded to your comments. I tried the reference path locally but I couldn't get that to work. I can redefine the type I need inside the file with:

`interface Requireable<T> extends React.Validator<T>{
  isRequired: React.Validator<T>
}` 

we could also switch the 'Requireable<T>' typing to an 'any' typing. Beyond that I'm out of my expertise

**Comment by bcherny** [comment]
@chrisgarber

> `interface Requireable<T> extends React.Validator<T>{ isRequired: React.Validator<T> }`

It's a little gross to redefine existing types, but let's do this so we don't end up pulling in extra code for some users.

**Comment by bcherny** [comment]
Published as 3.2.0.

**Pull request merged**